deepseg.cc

```diff
@@ -110,6 +131,111 @@ void *grab_thread(void *arg) {
 	return NULL;
 }
 
+void init_tensorflow(calcinfo_t *info) {
+	// Load model
+	info->model = tflite::FlatBufferModel::BuildFromFile(info->modelname);
+	TFLITE_MINIMAL_CHECK(info->model != nullptr);
+
+	// Build the interpreter
+	tflite::ops::builtin::BuiltinOpResolver resolver;
+	// custom op for Google Meet network
+	resolver.AddCustom("Convolution2DTransposeBias", mediapipe::tflite_operations::RegisterConvolution2DTransposeBias());
+	InterpreterBuilder builder(*info->model, resolver);
+	builder(&interpreter);
+	TFLITE_MINIMAL_CHECK(interpreter != nullptr);
+
+	// Allocate tensor buffers.
+	TFLITE_MINIMAL_CHECK(interpreter->AllocateTensors() == kTfLiteOk);
+
+	// set interpreter params
+	interpreter->SetNumThreads(info->threads);
+	interpreter->SetAllowFp16PrecisionForFp32(true);
+
+	// get input and output tensor as cv::Mat
+	info->input = getTensorMat(interpreter->inputs ()[0],info->debug);
+	info->output = getTensorMat(interpreter->outputs()[0],info->debug);
+	info->ratio = (float)info->input.cols/(float) info->input.rows;
+
+	// initialize mask and square ROI in center
+	info->roidim = cv::Rect((info->width-info->height/info->ratio)/2,0,info->height/info->ratio,info->height);
+	info->mask = cv::Mat::ones(info->height,info->width,CV_8UC1);
+	info->mroi = info->mask(info->roidim);
+
+	// erosion/dilation element
+	info->element = cv::getStructuringElement( cv::MORPH_RECT, cv::Size(5,5) );
+
+	// create Mat for small mask
+	info->ofinal = cv::Mat(info->output.rows,info->output.cols,CV_8UC1);
+}
+
+void calc_mask(calcinfo_t *info) {
+	// map ROI
+	cv::Mat roi = info->raw(info->roidim);
+
+	// resize ROI to input size
+	cv::Mat in_u8_yuv, in_u8_rgb;
+	cv::resize(roi,in_u8_yuv,cv::Size(info->input.cols,info->input.rows));
+	cv::cvtColor(in_u8_yuv,in_u8_rgb,CV_YUV2RGB_YUYV);
+	// TODO: can convert directly to float?
+
+	// convert to float and normalize values to [-1;1]
+	in_u8_rgb.convertTo(info->input,CV_32FC3,1.0/128.0,-1.0);
+
+	// Run inference
+	TFLITE_MINIMAL_CHECK(interpreter->Invoke() == kTfLiteOk);
+
+	float* tmp = (float*)info->output.data;
+	uint8_t* out = (uint8_t*)info->ofinal.data;
+
+	// find class with maximum probability
+	if (strstr(info->modelname,"deeplab"))
+	for (unsigned int n = 0; n < info->output.total(); n++) {
+		float maxval = -10000; int maxpos = 0;
+		for (int i = 0; i < cnum; i++) {
+			if (tmp[n*cnum+i] > maxval) {
+				maxval = tmp[n*cnum+i];
+				maxpos = i;
+			}
+		}
+		// set mask to 0 where class == person
+		uint8_t val = (maxpos==pers ? 0 : 255);
+		out[n] = (val & 0xE0) | (out[n] >> 3);
+	}
+
+	// threshold probability
+	if (strstr(info->modelname,"body-pix"))
+	for (unsigned int n = 0; n < info->output.total(); n++) {
+		// FIXME: hardcoded threshold
+		uint8_t val = (tmp[n] > 0.65 ? 0 : 255);
+		out[n] = (val & 0xE0) | (out[n] >> 3);
+	}
```

Comment:
Misleading indentation. Please use braces and indent the for loop.

Reply:
fixed